### PR TITLE
feat(dx): agent orchestra for frontend issue-to-PR workflow

### PR DESCRIPTION
**Pull request by zJaaal**
## Summary

This PR introduces an **AI agent orchestra** for the dotCMS frontend development workflow. It wires together a set of specialized Claude Code subagents, a skill (slash command), and a shell hook to automate the full journey from a GitHub issue to a reviewed, QA-verified draft PR — all from within a Claude Code session.

The system is **frontend-only** (scoped to `core-web/` — Angular, TypeScript, SCSS) and is invoked with a single command: `/issue-to-pr <issue-number>`.

---

### `.claude/skills/issue-to-pr/SKILL.md`

A Claude Code *skill* — the slash command `/issue-to-pr`. It is the orchestration script for the entire pipeline.

**Why it exists:** Without this skill, working a GitHub issue from end-to-end requires many manual steps: reading the issue, researching the codebase, designing pseudocode, writing code, reviewing it, running QA, and creating a PR. This skill coordinates specialized subagents to handle each concern in the right order, keeping the developer focused on decisions rather than plumbing.

**8-stage pipeline:**

```
Stage 1 — Resolve issue number
  Parse /issue-to-pr argument. Fall back to branch name (issue-NNNNN-*).

Stage 2 — Fetch issue + run parallel agents
  gh issue view → feeds two agents simultaneously:
    • dotcms-issue-validator   → checks completeness (SUFFICIENT / NEEDS_INFO)
    • dotcms-product-analyst   → product research: pseudocode, AC, edge cases, coder brief

Stage 3 — Interactive product Q&A
  Present product analysis. Ask open questions. Gather answers.
  Confirm final Coder Brief before coding begins.

Stage 4 — Create feature branch
  git checkout -b issue-<number>-<slug>  (if not already on one)

Stage 5 — Implementation
  Present Coder Brief + file list. Implement using pseudocode as spec.

Stage 6 — Review changed files
  git diff HEAD → filter core-web/ frontend files
  dotcms-file-classifier → review buckets
  Parallel reviewers: TypeScript / Angular / Tests / SCSS+HTML
  Consolidate by severity: 🔴 Critical → 🟡 Important → 🔵 Quality
  Stage 6d: fix → re-run tests → re-run QA → reply + resolve threads

Stage 7 — QA
  Derive positive / negative / edge-case scenarios from product analysis.
  Ask developer for real test data (host, API key, site ID, content types).
  Execute based on what changed:
    • SDK libraries  → build → npm pack → Node.js qa-test.mjs
    • Angular UI     → dev server + browser inspection
    • Services       → Jest tests + manual API calls
    • SCSS only      → visual screenshot comparison
  Block PR creation on any failure. Re-run after every review fix.

Stage 8 — Create draft PR + move issue
  github-workflow-manager creates draft PR (Closes #N, summary, AC checklist, QA results).
  If PR is not draft → move linked issue to "In Review" on project board.
```

**Review comment protocol (Stage 6d):** Reply to each comment before resolving it. End replies with `— 🤖 Claude` signature so the reviewer knows who responded and can continue the conversation from a clean spot.

---

### `.claude/agents/dotcms-product-analyst.md`

A Claude Code *custom subagent* — the "Product Analyst" role in Stage 2.

**Why it exists:** Generic AI tends to jump to solutions. This agent is constrained to a 5-phase research process: understand the issue → research existing behavior → map acceptance criteria → write pseudocode in real Angular/NgRx patterns → identify edge cases. It only asks questions that are genuinely blocking.

**Configuration:**
```yaml
model: sonnet       # Cost-balanced: good reasoning, lower latency than opus
color: purple
allowed-tools: [Grep, Glob, Read]
maxTurns: 20        # Enough for 5-phase research without runaway loops
```

**Research constraints:** max 3 Grep/Glob, max 4 Read calls — focused, not exhaustive. Pseudocode must use real patterns from the codebase (actual signal names, store methods, API endpoints). Open Questions must be genuinely blocking — if answerable from code or common sense, answer it directly in the Coder Brief.

---

### `.claude/hooks/post-commit-review.sh`

A shell script wired as a `PostToolUse` Claude Code hook. Fires after every `Bash` tool call.

**Why it exists:** Developers often commit without remembering to run the appropriate reviewers. This hook passively watches for successful frontend commits and nudges the developer — no pipeline invocation required.

**How it works:**
1. Receives Claude Code hook JSON on stdin (tool_input + tool_response)
2. Extracts bash command and output via Python JSON parsing
3. Guards: only acts on `git commit` commands with a successful commit in output (`[branch abc1234]`)
4. Gets frontend files from last commit via `git diff --name-only HEAD~1 HEAD`
5. If no frontend files → exits silently (no noise for backend-only commits)
6. Builds recommended reviewer list based on file types:
   - `.ts` → `dotcms-typescript-reviewer`
   - `.ts`/`.html` → `dotcms-angular-reviewer`
   - `.spec.ts` → `dotcms-test-reviewer`
   - `.scss`/`.css`/`.html` → `dotcms-scss-html-style-reviewer`
7. Outputs a one-line nudge to Claude's context

The hook is **informational only** — it does not auto-trigger reviewers. The developer decides whether to act.

**Why `PostToolUse` on `Bash` instead of a git hook:** Claude Code runs git via the Bash tool. Claude Code's hook system provides the full tool input + output in JSON, letting us inspect whether the commit actually succeeded without a separate git call.

---

### `.claude/settings.json` (modified)

Added `PostToolUse` hook registration:
```json
"hooks": {
  "PostToolUse": [
    { "matcher": "Bash", "hooks": [{ "type": "command", "command": ".claude/hooks/post-commit-review.sh" }] }
  ]
}
```

---

## Agents used in the pipeline

| Agent | Role | When |
|---|---|---|
| `dotcms-issue-validator` | Checks issue completeness | Stage 2 (parallel) |
| `dotcms-product-analyst` | Deep product research + pseudocode | Stage 2 (parallel) — **new** |
| `dotcms-file-classifier` | Sorts changed files into review buckets | Stage 6a |
| `dotcms-typescript-reviewer` | TypeScript type safety | Stage 6b |
| `dotcms-angular-reviewer` | Angular patterns | Stage 6b |
| `dotcms-test-reviewer` | Test quality | Stage 6b |
| `dotcms-scss-html-style-reviewer` | Styling standards | Stage 6b |
| `github-workflow-manager` | Creates the draft PR | Stage 8 |

All reviewer agents and `dotcms-issue-validator` / `dotcms-file-classifier` / `github-workflow-manager` are existing agents. `dotcms-product-analyst` is new.

---

## Test plan

- [ ] Run `/issue-to-pr <number>` on a real frontend issue end-to-end
- [ ] Verify Stage 2 agents run in parallel
- [ ] Verify Stage 3 Q&A waits for user answers before proceeding
- [ ] Verify Stage 6 launches only reviewers matching the changed file types
- [ ] Verify Stage 7 asks for test data and blocks on QA failure
- [ ] Commit a frontend file → verify post-commit hook fires with correct reviewer recommendation
- [ ] Commit a backend-only file → verify hook stays silent
- [ ] Verify Stage 8 moves the issue to "In Review" (requires `project` GitHub token scope: `gh auth refresh -s project`)

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Closes #34909

### CODE REVIEW

**Review by spbolton — APPROVED** [review]
**PR Requires a Linked Issue**

This PR breaks our rule requiring all PRs to have an associated issue — there's even a failing check because of it. That said, this is a separate concern from the main review.

If you don't already have an issue, you can create one easily using Claude with the new skill or by running `/create-issue`. I'm happy to share a doc explaining why all PRs should have linked issues, even for unplanned work outside a sprint.

**On the friction of creating issues**

Yes, it can feel like extra overhead to create an issue just to open a PR — but with AI tooling, this becomes a non-issue. Claude can enforce the rule and create both the PR and the issue together.

**On skill design**

For this PR, I think we can merge it and see how it goes. However, I'd like us to think about skills as covering an *area of responsibility* rather than mapping one-to-one with commands. A single skill should support multiple commands and options — otherwise we end up with duplication and fragmentation.

With that in mind:
- [`[dotcms-github-issues](https://github.com/dotCMS/core/tree/main/.claude/skills/dotcms-github-issues)`](https://github.com/dotCMS/core/tree/main/.claude/skills/dotcms-github-issues) already handles issue management broadly
- A companion skill for PRs should handle PR creation, updates, and linking
- Branch/worktree creation and the dev/test cycle are a separate concern that belongs upstream of the PR skill

If this gets merged, the follow-up would be a refactor to consolidate the core rules and flow — similar to how I refactored `create-issue` into `dotcms-github-issues` while keeping `/issue-to-pr` working as a command within it.

File: .claude/skills/fetch-issue/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+---
```
Comment (spbolton) [inline .claude/skills/fetch-issue/SKILL.md:1]: should we be using the dotcms-github-issues skill for interacting with the issues rather than creating a new skill.

**Review by spbolton — CHANGES_REQUESTED** [review]

File: .claude/skills/fetch-issue/SKILL.md (hunk 1 above)
Comment (zJaaal) [reply .claude/skills/fetch-issue/SKILL.md:1]: Good question — the separation is intentional. `dotcms-github-issues` is a user-facing conversational skill (mode detection, label inference, template selection, Q&A flows). `fetch-issue` is a silent pipeline utility: resolve a number from context, fetch JSON, check for existing branches/PRs, pass structured output to the next step. Pulling a 600-line conversational skill into the middle of an automated pipeline would reintroduce the context-bloat problem this decomposition was designed to avoid. Same principle as the class/method analogy @spbolton raised — `dotcms-github-issues` is the user-facing API, `fetch-issue` is the internal plumbing. — 🤖 Claude

File: .claude/skills/branch/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+```bash
+git branch --show-current
+```
+
+If already on an `issue-<number>-*` branch → skip to Step 3.
+
+If `existing_branch` was set by `/fetch-issue` and user chose `reuse` → check it out:
+
+```bash
+git checkout <existing_branch>
```
Comment (spbolton) [inline .claude/skills/branch/SKILL.md:44]: I am concerned we do not have any acknowledgement that we may be already working on an existing branch with committed and uncommitted changes.    It is likely to throw errors if there are uncommitted changes, but also may have problems putting the folder back into the previous state.    Note there is already scripting in dotcms-utilities for "git issue-branch" which uses "git smart-switch" that handles these behaviors.   We do not need to use these scripts directly, but I think the logic is worthy of using instead of creating new simpler logic that does not handle a lot of cases.   Maybe we point claude at the dotcms-utilties repo to analyze the logic and suggest how to improve the logic in here with it.

**Review by spbolton — DISMISSED** [review]

File: .claude/skills/fetch-issue/SKILL.md (hunk 1 above)
Comment (spbolton) [reply .claude/skills/fetch-issue/SKILL.md:1]: Good point. To be clear, the multi-mode design in dotcms-github-issues was intentional — the goal was to capture the full   
  behavior of "interacting with issues" as a single area of responsibility, with shared rules for label inference, team caching,
   and refinement that stay consistent across operations. That's a different priority than context optimization.
                                                                                                                                
  That said, looking at how you've structured the decomposition here — scoped allowed-tools per skill, thin orchestrator, shared
   reference files — there are real benefits we could apply back to dotcms-github-issues. The modes map naturally to separate
  skills with shared references, giving you both the context efficiency and the single source of truth for the rules.

  The key thing to protect as we go in this direction is the shared-references pattern you're already using with
  issue-refinement.md. That's what keeps decomposed skills from drifting apart.
